### DISCUSSION TITLE
The Linux Mint configuration interferes with Autokey

### DISCUSSION THREAD

**Discussion by ineuw** [autokey/autokey#766] [Q&A]
@Elliria and @josephj11, many thanks for the input and examples and I cleaned all my scripts accordingly.

The cleanup didn't correct the problem, regardless of the methods tried. So, configuring the Diodon clipboard viewer to track selection by mouse or keyboard navigation keys, revealed the problem.

It's the LM Windows focus option which messes everything up! Autokey, as well as TextPad, and Irfanview in Wine, are all equally affected. 

Also noticed that a mouse click does not always force a focus in other apps, and I can't figure out how to overcome this issue. 

LM windows settings, with their mouse options related to positioning, destroys the clipboard content of mouse selected text, and I get the "X selection is empty" error in Autokey. When text is selected by keyboard navigation keys, the macro always works.

LM Windows Behavior panel offers three possible settings, shown in the image, but only "Click" works intermittently! The other two options never.

I tried two Linux Mint flavors so far, and the problem is the same.

![230203_LM_window_and_mouse_options](https://user-images.githubusercontent.com/10367016/216630089-4bcb0b23-12ac-4008-bd70-eace77b49c5c.jpg)

**Comment by josephj11** [comment]
If things don't "click" for you, you can try using the `window.activate(title, switchDesktop=False, matchClass=False)` AutoKey API call.

If you need to get fancier than that, then call out to xdotool. It has this amazing command chaining feature  that lets you do several things in a row like find a window, activate it, and interact with it all in one command invocation.

It's way too complicated to describe in detail here and it can be annoying to debug. If you ever need to do it, we can do private emails or chat.

You can also look into AutoKey's Xautomation API to search the screen for objects to interact with. You can call its components directly to get a bunch of functionality that's not in our API. That's another extended discussion.

Problems like you are describing are in the domain of the DTE (desktop environment) or, occasionally, one level lower in the WM (window manager). You could try another DTE such as KDE and see if that helps. I haven't done anything at the WM level, but those can be swapped out as well.

**Reply by ineuw** [reply]
Thanks for your quick reply, the API call is clear. Did you ever come across this elsewhere? 

First, I want to exhaust Linux Mint offerings. The next up is Linux Mint Xfce. KDE and QT was tried many years ago, but I am sure it changed as well.

**Comment by josephj11** [comment]
No, but we all occasionally tussle with DTE's and WM's. The usual problem is when someone tries to use a hotkey combination that the DTE or WM has assigned and it gets and consumes the hotkey and AutoKey never sees it. When I installed kubuntu 22.04 on my other notebook, I discovered that KDE had claimed a whole raft of hotkeys for obscure things,

**Reply by ineuw** [reply]
Displaying my ignorance here, I never dared asking about application shortcuts and AK. I blamed my problems on X clipboard, GTK, and Gnome, because I assumed that there some sort of X controlled key source hierarchy.

My long running frustration with system hotkeys was partially resolved by changing all Autokey "ctrl" shortcuts to "super". I've done this after studying the trends in various applications' shortcuts. When using Wine apps, Autokey is unnecessary and is disabled.

Linux Mint Cinnamon, which is also based on Ubuntu, a number of keyboard assignments can be cleared manually, but it's not a centralized system. I asked in the LM forum if it's possible to disable them all programmatically, but received no reply.

**Reply by josephj11** [reply]
> My long running frustration with system hotkeys was partially resolved by changing all to keys. 

Don't know what you mean by this.

I use AutoKey all the time, but almost never in text editors.

As I'm sure you are aware, the AutoKey panel icon context menu includes an Enable monitoring checkbox. When unchecked, AutoKey will not detect any triggers, but will still be available to run actions from the panel context menu and from the main AutoKey window. You can also use window filters to restrict when triggers will be recognized.

Your last comment seems to contradict your earlier statement (on Gitter?) that you had no problems with AutoKey on LMC.

**Reply by ineuw** [reply]
I corrected the post. I use Autokey in the built Wikisource text editor and offline text editors. Firefox shortcuts are disabled programmatically, and the Vivaldi browser shortcuts can be disabled in preferences.

**Comment by Elliria** [comment]
Hey there. Is this [Linux Mint forum post](https://forums.linuxmint.com/viewtopic.php?t=312466) of any use? It can be a major hassle when focus is either stolen or not stolen depending on what you want to have happen. It's something I wrestled with for quite some time in Windows before they finally offered a way to manage it without third-party software, so I feel your pain, at least when it comes to my work computer. So far, Kubuntu is well-behaved as far as focus is concerned.

**Reply by ineuw** [reply]
Many thanks. It may be what I need, if I understand the settings correctly. Xfce4 controls may be different in looks but I think that underneath the hood, they are the same. The drop-down list of Windows focus mode are Click, Sloppy, and Mouse.

![230204_windows_focus_settings_in_LMC](https://user-images.githubusercontent.com/10367016/216771777-c68b265d-3699-40a5-86bd-86bfff53bf79.jpg)

**Reply by Elliria** [reply]
I'm not familiar with that particular configuration, but it looks like it will be able to do it. Your magic will be in the "Window focus mode" on that screen, I would think.

All I know is that I've always felt very strongly that wherever my mouse happens to be ought to take focus, but not everyone feels the same way.

**My preference:**
If two windows are open side-by-side and both can be scrolled by turning the mouse wheel:
* If you hover over **this** one and turn the wheel, **this** one should scroll.
* If you hover over **that** one and turn the wheel, **that** one should scroll.
* A mouse-click should **never** be required to change focus.

**Other folks' preference:**
If two windows are open side-by-side and both can be scrolled by turning the mouse wheel:
* If you hover over **this** one and turn the wheel, **this** one should scroll.
* If you hover over **that** one and turn the wheel, **this** one should scroll.
* A mouse-click should **always** be required to change focus.

As you can imagine, either type of user would be very disoriented and disgruntled if they had to deal with the opposite kind of behavior.

Interestingly, some operating systems don't agree with my preference. I'm fine with that as long as there's an option for me to go in and change it if I like. I used to have to use a third-party program in Windows to control that, but they've finally given us a configuration setting that I immediately use on any new installation and that I'm very grateful for since many employers insist that we use Windows. The Ubuntu flavors seem to all offer a way to control this behavior, but I haven't checked them all. I've got my preferred behavior in Kubuntu. It looks like you'll be able to control it in Linux Mint, too.

**Reply by ineuw** [reply]
Hi, before replying to the posts of the icon thread, I wrote a long reply last night for this post and lost it. Since the image I first posted, an additional setting on that panel needed to be disabled because it interfered with Autokey. Then, everything worked perfectly with the exception below. 

There is a difference for X-clipboard if "ctrl+a" (select all) as opposed to the mouse or keyboard select. When using "ctrl+a", I must use @josephj11's API call to get the scripts working. The article which best explained all aspects of the X-clipboard are misplaced temporarily but is in the process of being tracked down. I must add that to our Help pages.

**Comment by Elliria** [comment]
I think the X-clipboard article is the one that's linked to in the [X Clipboards](https://github.com/autokey/autokey/wiki/More-than-you-ever-wanted-to-know-about-X-clipboards) AutoKey wiki page.

**Reply by ineuw** [reply]
Unfortunately, this isn't it. I also looked at this an hour ago. :-). But will find it because I came across it when searching, and noticed that it was already bookmarked.

**Reply by Elliria** [reply]
Well, darn. I hope you find it and look forward to seeing it when you do. As @josephj11 said, the clipboard is a bit of a mystery, so the more information we can get, the better.

**Comment by josephj11** [comment]
Several unrelated comments:

AFAIK, there are no standard or rules about the use and assignment of hotkeys in Linux. It would be amazing to have an app that would catalog them all and tell you if one you want to use is already in use somewhere else, but I don't think that exists and it might be very difficult or almost impossible to write one. Maybe AI could help. If there were such a standard it would probably live at https://www.freedesktop.org/.

Losing unsubmitted posts is a real PITA. Take a look at the Typio Form Recovery add-on for browsers. It's a real lifesaver!

When you find that clipboard article, please add it to our wiki where the current clipboard article is or link to it. Clipboards are another typical computer thing that looks very simple, but is anything but when you actually get into the details.

I think I fixed our wiki, but in case I missed a link or you have an old bookmark, be sure to look [here](Autokey.github.io/autokey) for our current documentation - notably the API. This one is generated directly from the source code and will automatically include any changes made in the future provided it has been regenerated. I'm not sure what, if anything, triggers its regeneration automatically.

**Comment by josephj11** [comment]
Several unrelated comments:

AFAIK, there are no standard or rules about the use and assignment of hotkeys in Linux. It would be amazing to have an app that would catalog them all and tell you if one you want to use is already in use somewhere else, but I don't think that exists and it might be very difficult or almost impossible to write one. Maybe AI could help. If there were such a standard it would probably live at https://www.freedesktop.org/.

Losing unsubmitted posts is a real PITA. Take a look at the Typio Form Recovery add-on for browsers. It's a real lifesaver!

When you find that clipboard article, please add it to our wiki where the current clipboard article is or link to it. Clipboards are another typical computer thing that looks very simple, but is anything but when you actually get into the details.

I think I fixed our wiki, but in case I missed a link or you have an old bookmark, be sure to look [here](Autokey.github.io/autokey) for our current documentation - notably the API. This one is generated directly from the source code and will automatically include any changes made in the future provided it has been regenerated. I'm not sure what, if anything, triggers its regeneration automatically.

**Reply by ineuw** [reply]
Thanks for the info and the new webpage is beautiful!!!

**Reply by josephj11** [reply]
Yeah. It's nice. Either Bluedrink9 or Sam did that around a year ago. I finally updated the links. The old docs were also generated, but apparently, that program/process broke and they replaced it with something much better. It includes the new API calls that were added in 0.96.0 and also the Xautomation stuff that was isolated off in the new features page before.

**Comment by josephj11** [comment]
Several unrelated comments:

AFAIK, there are no standard or rules about the use and assignment of hotkeys in Linux. It would be amazing to have an app that would catalog them all and tell you if one you want to use is already in use somewhere else, but I don't think that exists and it might be very difficult or almost impossible to write one. Maybe AI could help. If there were such a standard it would probably live at https://www.freedesktop.org/.

Losing unsubmitted posts is a real PITA. Take a look at the Typio Form Recovery add-on for browsers. It's a real lifesaver!

When you find that clipboard article, please add it to our wiki where the current clipboard article is or link to it. Clipboards are another typical computer thing that looks very simple, but is anything but when you actually get into the details.

I think I fixed our wiki, but in case I missed a link or you have an old bookmark, be sure to look [here](https://Autokey.github.io/autokey) for our current documentation - notably the API. This one is generated directly from the source code and will automatically include any changes made in the future provided it has been regenerated. I'm not sure what, if anything, triggers its regeneration automatically.

**Comment by Elliria** [comment]
I'm getting a 404 with the link to the new documentation.

**Reply by josephj11** [reply]
Oops. Fixed. It was GitHub's version of autocorrect. If you insert a link without the protocol header (https:// etc.), it adds its own idea of what it thinks should be there which for me is never right.

**Reply by Elliria** [reply]
That's working perfectly now, thanks.

**Comment by ineuw** [comment]
So did I but it's fixed since :-)
